update preferences interface and add tests

rishichawda opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail, what problems does it solve? -->

Updated the preferences interface to provide generic methods for reading and writing preferences

- `::get` - Depending on whether the username and hostname are known, if either calls `CFPreferencesCopyAppValue` or `CFPreferencesCopyValue`. Will return value if it exists, otherwise nil.
- `::get!` - Calls `::get` and if no value was returned, raise `PreferenceDoesNotExist` error
- `::set` - Depending on whether the username and hostname are known, if either calls `CFPreferencesSetAppValue` or `CFPreferencesSetValue`, followed by a `CFPreferencesAppSynchronize` call. Returns `true` if operation is successful, `false` otherwise.
- `::set!` - Call `::set` and raise `PreferenceSyncFailed` if `false` is returned.

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

rishichawda committed:
feat(preferences.rb): update preferences interface

Signed-off-by: rishichawda <rishichawda@users.noreply.github.com>

rishichawda committed:
test(preferences): update tests

Signed-off-by: rishichawda <rishichawda@users.noreply.github.com>

chef-expeditor[bot] committed:
Updating references of master to main as part of Expeditor agent rename.

logwolvy commented on lib/corefoundation/exceptions.rb:
```diff
@@ -0,0 +1,16 @@
+  module Exceptions
+    # Raised when the preference value couldn't be found for a domain.
+    class PreferenceDoesNotExist < StandardError
+      def initialize(key, domain, hostname)
```
`hostname` should be optional

rishichawda replied:
Yep. Been busy for a while so couldn't get back to the code. Updating this now.

logwolvy commented on lib/corefoundation/preferences.rb:
```diff
@@ -13,55 +10,120 @@ module CF
-        plist_ref = if username && hostname
-                      CF.CFPreferencesCopyValue(key.to_cf, application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
-                    else
+        plist_ref = if username.nil? or hostname.nil?
```
`if username && hostname` is cleaner

logwolvy replied:
status on these?

logwolvy commented on lib/corefoundation/preferences.rb:
```diff
@@ -13,55 +10,120 @@ module CF
       def get!(key, application_id, username = nil, hostname = nil)
-        get(key, application_id, username, hostname) || (raise CF::Error, "Preference key `#{key}` not found")
+        value = get(key, application_id, username, hostname)
+        if value.nil?
```
```ruby
get(key, application_id, username, hostname) || raise(CF::Exceptions::PreferenceDoesNotExist.new(key, application_id, hostname))
```
was cleaner, we can keep that

rishichawda replied:
What you are suggesting will raise an error for booleans too. We would only want to raise it if it's nil.

logwolvy replied:
If that function could potentially return boolean, it shouldn't. It should only return either the value or nil

rishichawda replied:
domain properties can have boolean values.

logwolvy commented on lib/corefoundation/preferences.rb:
```diff
@@ -13,55 +10,120 @@ module CF
+      # @return [TrueClass, FalseClass] Returns true if preference was successfully written to storage, otherwise false.
+      #
+      def set(key, value, application_id, username = nil, hostname = nil)
+        if username.nil? or hostname.nil?
```
Same here

logwolvy commented on lib/corefoundation/preferences.rb:
```diff
@@ -13,55 +10,120 @@ module CF
+      # @return [VALUE] Returns nil if preference value is successfully written.
+      #
+      def set!(key, value, application_id, username = nil, hostname = nil)
+        raise(CF::Exceptions::PreferenceSyncFailed.new(key, application_id, hostname)) unless set(key, value, application_id, username, hostname)
```
here also, we can just use `||`

rishichawda replied:
Doesn't have any functional difference. unless is more idiomatic.

logwolvy replied:
shortcircuit evaluation is different from conditionals on multiple levels but even if you want to continue using condition I recommend 2 things -
1. Atleast make this and that -> https://github.com/chef/corefoundation/pull/21/files#diff-7a48df7397684c3a72b5e2e4bac4b585f806d0a0bb9d309eaecd54c1e06dd932L44 consistent
2. keep `action` before the `error` for readability, i.e. keep `set(key, value, application_id, username, hostname)` before `raise`

rishichawda replied:
1. `get` won't be using `||` because boolean is a valid type for domain properties. 
2. Trying to move the `set` call to left would make it a `if not` statement which according to the style guide should be replaced with `unless`.  

So let's go with what we have here.

logwolvy replied:
shortcircuit evaluation/control flow was good enough, https://github.com/rubocop/ruby-style-guide#if-as-a-modifier but since you have already changed, let's keep this

rishichawda committed:
make hostname optional

Signed-off-by: rishichawda <rishichawda@users.noreply.github.com>

rishichawda commented:
Updated the PR with review changes and rebased. This should wait for https://github.com/chef/corefoundation/pull/15 to be fully updated and merged to master.

rishichawda commented:
Needs to wait for #16

tas50 committed:
Merge pull request #16 from chef/update-ruby-gem

Cleanup and modernize the gem

chef-ci committed:
Bump version to 0.3.2 by Chef Expeditor

Obvious fix; these changes are the result of automation not creative thinking.

dependabot[bot] committed:
Update chefstyle requirement from = 2.0.7 to = 2.0.9

Updates the requirements on [chefstyle](https://github.com/chef/chefstyle) to permit the latest version.
- [Release notes](https://github.com/chef/chefstyle/releases)
- [Changelog](https://github.com/chef/chefstyle/blob/main/CHANGELOG.md)
- [Commits](https://github.com/chef/chefstyle/compare/v2.0.7...v2.0.9)

---
updated-dependencies:
- dependency-name: chefstyle
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

tas50 committed:
Merge pull request #24 from chef/dependabot/bundler/chefstyle-eq-2.0.9

Update chefstyle requirement from = 2.0.7 to = 2.0.9

chef-ci committed:
Bump version to 0.3.3 by Chef Expeditor

Obvious fix; these changes are the result of automation not creative thinking.

logwolvy commented on spec/preferences_spec.rb:
```diff
@@ -0,0 +1,67 @@
+  describe "self.set" do
+    context "when called with valid domain/default pair" do
+      it "writes correct value and returns true" do
+        expect(CF::Preferences.set(@validKey, @value, @validDomain)).to eq true
```
All these tests touch/modify OS/filesystem where these are run. Even though you set the values back to nil but there will always be side-effects like these (and more). So, could you mock out and implement these unit tests?

rishichawda replied:
mocking out the calls wasn't actually testing anything so I had to do this. Since I'm using a fake domain, we don't need to worry about it

rishichawda replied:
example: https://github.com/chef/chef/blob/main/spec/unit/resource/macos_user_defaults_spec.rb

logwolvy replied:
> mocking out the calls wasn't actually testing anything so I had to do this. Since I'm using a fake domain, we don't need to worry about it

It depends on what you are mocking, there is plenty to unit test here and if there's nothing, then lose this test altogether (i.e. bad) but we certainly can't have it in this form. Doesn't matter if you use a fake domain, the tests interact with underlying OS's UserDefaults, and that's a side effect.

> example: https://github.com/chef/chef/blob/main/spec/unit/resource/macos_user_defaults_spec.rb

This is not testing a resource outcome but the utility methods `defaults_export_cmd` and `defaults_modify_cmd`, hence not interacting/mutating UserDefaults.

rishichawda committed:
review updates

Signed-off-by: rishichawda <rishichawda@users.noreply.github.com>

rishichawda commented:
All updates have been pushed and branch has been rebased on `main`

rishichawda committed:
fix missing refinements in test

Signed-off-by: rishichawda <rishichawda@users.noreply.github.com>
